agent/engine/docker_task_engine.go

```diff
@@ -602,41 +611,119 @@ func (engine *DockerTaskEngine) startContainer(task *api.Task, container *api.Co
+			Error:    NetworkNamespaceError{errors.Wrap(err, "provisionContainerResources: setup network namespace error")},
+		}
+	}
+
+	return DockerContainerMetadata{
+		DockerID: cniConfig.ContainerID,
+	}
+}
+
+// cleanupPauseContainer will clean up the network namespace of pause container
+func (engine *DockerTaskEngine) cleanupPauseContainer(task *api.Task, container *api.Container) error {
+	seelog.Infof("Task [%s]: Cleaning up the network namespace", task.String())
+
+	cniConfig, err := engine.BuildCNIConfigFromTaskContainer(task, container)
+	if err != nil {
+		return errors.Wrapf(err, "engine: failed cleanup task network namespace, task: %s", task.String())
+	}
+
+	return engine.cniClient.CleanupNS(cniConfig)
+}
+
+func (engine *DockerTaskEngine) BuildCNIConfigFromTaskContainer(task *api.Task, container *api.Container) (*ecscni.Config, error) {
+	cfg := &ecscni.Config{}
+
+	eni, err := task.GetTaskEni()
+	if err != nil {
+		return nil, err
+	}
+	cfg.ENIID = eni.ID
+
+	// Get the pid of container
+	containerInspectOutput, err := engine.client.InspectContainer(container.Name, inspectContainerTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.ContainerPID = strconv.Itoa(containerInspectOutput.State.Pid)
+	cfg.ContainerID = containerInspectOutput.ID
+
+	// TODO Confirm if there is better option instead of task arn
+	cfg.ID = task.Arn
+	cfg.ENIMACAddress = eni.MacAddress
+	// Get the primary ip of the eni
+	for _, ipv4 := range eni.IPV4Addresses {
+		if ipv4.Primary {
+			cfg.ENIIPV4Address = ipv4.Address
+			break
+		}
+	}
+	if cfg.ENIIPV4Address == "" {
+		return nil, errors.Errorf("engine: No primary ipv4 address found for this eni: %v", eni)
+	}
 
-	// TODO: Invoke libcni here
+	if len(eni.IPV6Addresses) != 1 {
+		return nil, errors.Errorf("engine: IPV6Addresses associated with eni isn't expected, eni:%v", eni)
+	}
+	cfg.ENIIPV6Address = eni.IPV6Addresses[0].Address
 
-	return DockerContainerMetadata{}
+	return cfg, nil
 }
 
 func (engine *DockerTaskEngine) stopContainer(task *api.Task, container *api.Container) DockerContainerMetadata {
-	log.Info("Stopping container", "task", task, "container", container)
+	seelog.Info("Stopping container, container: %v, task: %v", container, task)
 	containerMap, ok := engine.state.ContainerMapByArn(task.Arn)
 	if !ok {
 		return DockerContainerMetadata{
-			Error: CannotStopContainerError{fmt.Errorf("Container belongs to unrecognized task %s", task.Arn)},
+			Error: CannotStopContainerError{errors.Errorf("Container belongs to unrecognized task %s", task.Arn)},
 		}
 	}
 
 	dockerContainer, ok := containerMap[container.Name]
 	if !ok {
 		return DockerContainerMetadata{
-			Error: CannotStopContainerError{fmt.Errorf("Container not recorded as created")},
+			Error: CannotStopContainerError{errors.Errorf("Container not recorded as created")},
+		}
+	}
+
+	// Cleanup the pause container network namespace before stop the container
+	if container.IsInternal() && container.Name == api.PauseContainerName {
```

Comment:
Per the changes in https://github.com/aws/amazon-ecs-agent/pull/765, there's a proper type for the internal container now.  Please use that instead of the name.